extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,326 @@
+import {assert} from '../../../src/asserts';
+import {cidFor} from '../../../src/cid';
+import {getService} from '../../../src/service';
+import {log} from '../../../src/log';
+import {vsyncFor} from '../../../src/vsync';
+import {xhrFor, createXhrRequest} from '../../../src/xhr';
+
+
+const TAG_ = 'AmpUserNotification';
+
+/**
+ * @typedef {{
+ *   elementId: string,
+ *   ampUserId: string
+ * }}
+ */
+let GetRequestMetadata;
+
+/**
+ * @typedef {{
+ *   showNotification: boolean
+ * }}
+ */
+let GetResponseMetadata;
+
+/**
+ * @typedef {{
+ *   promise: (!Promise),
+ *   resolve: (function(*))
+ * }}
+ */
+let UserNotificationDefer;
+
+/**
+ * Component class that handles a simple notification.
+ */
+export class AmpUserNotification extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout != Layout.FILL;
+  }
+
+  /** @override */
+  buildCallback() {
+
+    this.element.classList.add('amp-user-notification');
+
+    /** @private @const {!Window} */
+    this.win_ = this.getWin();
+
+    /** @private @const {!UserNotification} */
+    this.userNotification_ = getUserNotification_(this.win_);
+
+    /** @private @const {!Vsync} */
+    this.vsync_ = vsyncFor(this.win_);
+
+    /** @private {boolean} */
+    this.activated_ = false;
+
+    assert(this.element.hasAttribute('id'), 'amp-user-notification should ' +
+        'have an id.');
+    /** @private {string} */
+
+    this.elementId_ = this.element.getAttribute('id');
+
+    /** @private {?string} */
+    this.ampUserId_ = null;
+
+    assert(this.element.hasAttribute('data-show-if-href'),
+        'amp-user-notification should have `data-show-if-href` attribute.');
+
+    /** @private @const {string} */
+    this.showIfHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-show-if-href'), this.element);
+
+    assert(this.element.hasAttribute('data-dismiss-href'),
+        'amp-user-notification should have `data-dismiss-href` attribute.');
+
+    /** @private @const {string} */
+    this.dismissHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-dismiss-href'), this.element);
+
+    this.options_ = {
+      credentials: this.element.getAttribute('credentials')
+    };
+
+    /** @private {?function(*)} */
+    this.resolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.dialogPromise_ = new Promise((resolve, reject) => {
+      this.resolve_ = resolve;
+    });
+
+    this.userNotification_.registerUserNotification(this.elementId_,
+        this.dialogPromise_);
+
+    this.registerAction('dismiss', this.dismiss.bind(this));
+  }
+
+  /** @override */
+  layoutCallback(visible) {
+    if (!this.activated_) {
+      this.activated_ = true;
+      return this.getAsyncCid_()
+          .then(this.fetchShowEndpoint_.bind(this))
+          .then(this.onFetchShowEndpointSuccess_.bind(this),
+              this.onRejection_.bind(this));
+    }
+    return Promise.resolve();
```

Comment:
How about setting `#activated_` to the promise above? That way if a second `#layoutCallback` happens before the requests come back, it returns the pending promise and not this already resolved one.